---
title: "Fig 1A-B: Analysis of the Haemopedia Database"
author: 
- name: Jason Cosgrove (jason.cosgrove@curie.fr)
  affiliation:  Institut Curie, UMR168
editor_options:
  chunk_output_type: inline
output:
  html_document:
    df_print: paged
    number_sections: yes
    toc: yes
  pdf_document:
    toc: yes
---
---

<style type="text/css">

body{ /* Normal  */
      font-size: 18px;
  }
td {  /* Table  */
  font-size: 8px;
}
h1.title {
  font-size: 38px;
  color: DarkRed;
}
h1 { /* Header 1 */
  font-size: 28px;
  color: DarkBlue;
}
h2 { /* Header 2 */
    font-size: 22px;
  color: DarkBlue;
}
h3 { /* Header 3 */
  font-size: 18px;
  font-family: "Times New Roman", Times, serif;
  color: DarkBlue;
}
code.r{ /* Code block */
    font-size: 10px;
}
pre { /* Code block - determines code spacing between lines */
    font-size: 14px;
}
</style>



# Introduction 
In this analysis we analyse the haemopedia bulk RNAseq database to identify metabolic signature of lineage bias. More specifically, we perform data QC and normalisation followed by differential expression and pathway enrichment analysis.



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Users/jasoncosgrove/Dropbox (Team_Perie)/Jason/Experiments/Dry_Lab/JCB4_metabolism_bioinformatics_pipeline')
#rm(list=ls())


library(Seurat)
library(dplyr)
library(edgeR)
library(limma)
#library(Glimma)
require(DESeq2)
library(cowplot)
library(ggplot2)
library(ggrepel)

source("/Users/jasoncosgrove/Dropbox (Team_Perie)/Jason/manuscripts/metabolism/July_2022/Raw_data/Figure1/1A_helpermethods.R")
load("/Users/jasoncosgrove/Dropbox (Team_Perie)/Jason/Experiments/Dry_Lab/JCB4_metabolism_bioinformatics_pipeline/genesets/geneset_Robjects/metabolic_signatures_metafate.Rda")

#set the seed so we can reproduce the results
set.seed(12345)


```


# Data QC and normalisation using the LIMMA package. 

Filter the dataset and normalise using limmas voom function. The voom method transforms count data to log2-counts per million (logCPM), then estimates the mean-variance relationship in the data to compute appropriate observation-level weights.


```{r load_and_normalise_data, echo=FALSE, eval= TRUE}

#load in the haemopedia dataset which we have stored as a DGEobject, this has been downloaded. 
load("datasets/Haemopedia/DGE_object_haemopedia.Rda")



# filter out activated lymphocytes, megs and pDCs as these are not relevant to our research  question
cells.to.remove <- c("GCB.1","GCB.2","SplPlsB.1" ,
                      "SplPlsC.1","SplPlsC.2",
                      "SplPlsC.3","BMPlsC.1","MemCD4T.1",
                      "EffCD4T.1","pDC.1",
                      "pDC.2",
                      "EffCD4T.2","MemCD4T.2",  
                      "MemCD8T.2","MemCD8T.1",
                      "MemCD4T.1", "pDC.1", "pDC.2",
                      "MegTPO.1",  "MegTPO.2")


#filter out lowly expressed genes, and also cells that we are not interested in. 
haemopedia.filtered <- filterData(haemopedia,cells.to.remove)

#add the lineage annotations for our analysis
lin <- performLineageAnnotation(haemopedia.filtered)


condition <-factor(lin)
design <- model.matrix( ~ 0 + condition)
colnames(design) <- levels( condition )

#calculate normalisation factors for each cell type
y <- calcNormFactors(haemopedia.filtered) 

#run the zoon transformation
v <- voom(haemopedia.filtered,design,plot = TRUE, span = 0.1,normalize.method = "quantile") 
#plot the library sizes for each sample as a sanity check
boxplot(v$E)


```


# Dimenstionality Reduction of Metabolic State

Do a PCA on the data using enzymes and transporters from the top 1000 most highly variable genes. 


```{r PCA}

#Create a DESEq project
dds <- DESeqDataSetFromMatrix(countData=y$counts,
                              DataFrame(condition=condition),
                              design=~ 0 + condition)
dds <- estimateSizeFactors(dds)

### For exploratory analysis, let's normalize by VST
rld <- vst(dds, blind=TRUE)
d.rld <- assay(rld)

#select metabolically associated genes within the top 1000 most variably expressed genes
var_genes <- rowVars(d.rld)
select_var <- order(var_genes, decreasing=TRUE)[1:1000]
genes <- intersect(rownames(y)[select_var], c(metabolic.signatures$justMetabolicKegg,metabolic.signatures$justTransport))

#now run a PCA on just those genes
pca <- prcomp(t(d.rld[genes,]))
df <- as.data.frame(pca$x)
df$condition <- condition
percentVar <- round(pca$sdev^2/sum(pca$sdev^2)*100)



#tiff("PCA.tiff", units = "in" , width = 10, height = 10 , res = 300)
  print(p1.1 <- ggplot(df) + geom_point(aes(PC1, PC2, color=condition), size=3) +
  labs(title="", x=paste0("PC1: ",percentVar[1],"% variance"),
       y=paste0("PC2: ",percentVar[2],"% variance")) +
        theme_classic(base_size = 16) + theme(legend.title=element_blank()) + scale_color_manual(values=c("firebrick2","#E69F00","grey", "#56B4E9","grey")))
#dev.off()




```




# Differential expression analysis
Diffrential expression testing using Limma. Lets visualise our differentially expressed genes using a volcano plot, overlaying the expression of metabolic genes.

```{r degs}

#specify the contrast matrix that we want to use for our analysis
cont.matrix <- makeContrasts( E = Erythroid - (Myeloid + Lymphoid) / 2 ,
                              M = Myeloid - (Erythroid +Lymphoid)  / 2 ,
                              L = Lymphoid - (Myeloid + Erythroid) / 2 ,
                              levels = design )

#Now that we have the voom transformed data we can use limma to test for differential expression. First we fit a linear model for each gene using the lmFit function in limma. lmFit needs the voom object and the design matrix that we have already specified, which is stored within the voom object.
fit <- lmFit(v)

#Now we can apply the contrasts matrix to the fit object to get the statistics and estimated parameters of our comparison that we are interested in. 
fit.cont <- contrasts.fit(fit, cont.matrix)

#The final step is to call the eBayes function, which performs empirical Bayes shrinkage on the variances, and estimates moderated t-statistics and the associated p-values.
fit.cont <- eBayes(fit.cont, robust = T, proportion = 0.4)

#print a summary of the results
summa.fit <- decideTests(fit.cont)
summary(summa.fit)

#set the contrast names for when we plot the results
contrast.names <- c("E","M","L")

# get all metabolic genes
metabolic.genes <- c(metabolic.signatures$allGenesTested)




# custom volcano plot of limma differentially expressed genes
Volcano2 <- function(res, main="", fct=2, pt=0.05, xlim1 = -14,xlim2 = 14,gene.list){
  res$sign <- 0
  res$sign[which(res$adj.P.Val < pt & res$logFC> fct)] <- 1
  res$sign[which(res$adj.P.Val < pt & res$logFC < -fct)] <- 4
  
  #res$sign[which(res$adj.P.Val < pt & abs(res$logFC) > fct & res$genes %in% metabolic.genes)] <- 1
  p <- ggplot(data=res, aes(x=logFC, y=-log10(adj.P.Val), colour=as.factor(sign))) + geom_point( size=2) +
    
    scale_color_manual(name="", values=c("4" = "grey50","3" = "blue" ,"2" = "red","1"=
                                           rgb(67/255,138/255,201/255,0.4), "0"=rgb(220/255,220/255, 220/255,0.2))) +  
    ggtitle(paste0("Volcano Plot - Limma ", main)) +
    theme(legend.position = "none") + xlim(xlim1,xlim2) + 
    xlab("log2 fold change") + ylab("-log10 adj pvalue") + 
    geom_vline(xintercept=c(-fct, fct), linetype=2,colour = "grey",size = 1) + 
    geom_hline(yintercept=-log10(pt), linetype=2,colour = "grey",size = 1) +geom_text_repel(data=res[gene.list,],aes(label=genes),fontface = "bold",color = "black",min.segment.length = unit(0, 'lines'),size = 6)
  
  
  p <- p  + theme(
    text = element_text(size = 20),
    # Remove panel border
    panel.border = element_blank(),  
    # Remove panel grid lines
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    # Remove panel background
    panel.background = element_blank(),
    # Add axis line
    axis.line = element_line(colour = "grey")
  ) 
  
  
  p
  
  
  
}



LimmaVolcano <- function(res, main="", fct=2, pt=0.05){
  res$sign <- 0
  res$sign[which(res$adj.P.Val < pt & abs(res$logFC) > fct)] <- 1
  
  res$sign[which(res$adj.P.Val < pt & res$logFC > fct & res$genes %in% metabolic.genes)] <- 3
  
    res$sign[which(res$adj.P.Val < pt & res$logFC < -fct & res$genes %in% metabolic.genes)] <- 2
    
  p <- ggplot(data=res, aes(x=logFC, y=-log10(adj.P.Val), colour=as.factor(sign))) + geom_point( size=2) +
    
    scale_color_manual(name="", values=c("4" = "orange","3" = "blue" ,"2" = "red","1"=
                                           rgb(80/255,80/255, 80/255,0.2), "0"=rgb(220/255,220/255, 220/255,0.2))) +  
    ggtitle(paste0("Volcano Plot - Limma ", main)) +
    theme(legend.position = "none") + xlim(-12,12) + 
    xlab("log2 fold change") + ylab("-log10 adj pvalue") + 
    geom_vline(xintercept=c(-fct, fct), linetype=2) + 
    geom_hline(yintercept=-log10(pt), linetype=2)
  p
}
 

# do volcano plots for each lineage, overlaying the expression of all metabolic genes
res.E <- topTable(fit.cont, coef=1, number=nrow(v$E), adjust.method="BH")
res.E <- res.E[order(res.E$adj.P.Val),]
nsign.E <- length(which(res.E$adj.P.Val < 0.05))
nsignfc.E <- length(which(res.E$adj.P.Val < 0.05 & abs(res.E$logFC) > 1))



tiff(paste('/Users/jasoncosgrove/Desktop/',"E_volcano.tiff",sep = ""), width = 4, 
     height = 4, units = 'in', res = 300)                        
LimmaVolcano(res.E)
dev.off()

res.E.filtered <- res.E[res.E$adj.P.Val < 0.05,]

intersect(res.E.filtered$genes,metabolic.signatures$sequential_erythroid)


res.M <- topTable(fit.cont, coef=2, number=nrow(v$E), adjust.method="BH")
res.M <- res.M[order(res.M$adj.P.Val),]
nsign.M <- length(which(res.M$adj.P.Val < 0.05))
nsignfc.M <- length(which(res.M$adj.P.Val < 0.05 & abs(res.M$logFC) > 1))


#tiff(paste('Results/Module1/',"M_volcano.tiff",sep = ""), width = 5, 
#     height = 4, units = 'in', res = 300) 
LimmaVolcano(res.M,xlim1 = -10,xlim2 = 10)






#dev.off()
res.M.filtered <- res.M[res.M$adj.P.Val < 0.05,]
res.M.filtered <- res.M.filtered[order(res.M.filtered$logFC),]

intersect(res.M.filtered[res.M.filtered$logFC < -2,]$genes,metabolic.signatures$justMetabolicKegg)

intersect(res.M.filtered[res.M.filtered$logFC > 2,]$genes,metabolic.signatures$justMetabolicKegg)


gene.list <- c("G6pdx","Idh1","Hk3","Hexa","Gsr","Pkm","Agpat2","Pgam1","Atp8b4",
               'Slc14a1',"Odc1" ,    "Slc11a2",  "Cpox", "Slc16a1",  "Abcb4"  ,  "Aqp1" ,"Sptlc2","Hk2","Taldo1","Aldh1a1","Pfkm","Acacb","Pklr","Bpgm","Gpi1","Ldlr","Dgat1","Ido1"  ,   "Galc"   ,  "Miox"    , "Sat1" ,    "Renbp"  ,"Aldh3b1" , "Mboat1" ,  "Abca9" ,   "Acpp"  ,   "Cbr3" ,      "Chdh"  ,"Neu1","Atp6ap2")

tiff(paste('Results/Module1/',"M_volcano.tiff",sep = ""), width = 10, 
     height = 8, units = 'in', res = 300)   
Volcano2(res.M,xlim1 = -9,xlim2 = 9,gene.list = gene.list,fct = 1.5)

dev.off()

res.L <- topTable(fit.cont, coef=3, number=nrow(v$E), adjust.method="BH")
res.L <- res.L[order(res.L$adj.P.Val),]
nsign.L <- length(which(res.L$adj.P.Val < 0.05))
nsignfc.L <- length(which(res.L$adj.P.Val < 0.05 & abs(res.L$logFC) > 1))


#tiff(paste('Results/Module1/',"L_volcano.tiff",sep = ""), width = 5, 
#     height = 4, units = 'in', res = 300) 
LimmaVolcano(res.L,xlim1 = -12,xlim2 = 12)
#dev.off()

res.L.filtered <- res.L[res.L$adj.P.Val < 0.05,]

write.csv(res.L.filtered, "Results/Module1/DEGs_L.csv")
write.csv(res.E.filtered, "Results/Module1/DEGs_E.csv")
write.csv(res.M.filtered, "Results/Module1/DEGs_M.csv")




```


# Gene-set overrepresentation analysis

Geneset overrepresentation analysis: This works by seeing if your differentially expressed genes occur at a high frequency in a given geneset, relative to chance. The Null hypothesis here is that if a geneset represents 1% of all genes in the experiment then 1% of the genes in the pathway would be differentially expressed.   We perform these analyses using the kegga and goana methods available in the limma package. 


```{r geneset, eval = FALSE}

for(i in 1:length(contrast.names)){
  
  go <- goana(fit.cont, coef=i,species = "Mm",geneid = haemopedia$entrezgene[rownames(haemopedia$counts) %in% rownames(v$E)])
  kegg <- kegga(fit.cont, coef=i,species = "Mm",geneid = haemopedia$entrezgene[rownames(haemopedia$counts) %in% rownames(v$E)])
  
  go$FDR.up <- p.adjust(go$P.Up, method = "BH", n = length(go$P.Up))
  go$FDR.down <- p.adjust(go$P.Down, method = "BH", n = length(go$P.Down))
  
  kegg$FDR.up <-  p.adjust(kegg$P.Up, method = "BH", n = length(kegg$P.Up))
  kegg$FDR.down <- p.adjust(kegg$P.Down, method = "BH", n = length(kegg$P.Down))


  write.csv(go,   paste( "Results/Module1/" ,contrast.names[i],"GOANA_GO_pathways.csv",sep="_"))
  write.csv(kegg, paste( "Results/Module1/" ,contrast.names[i],"KEGGA_KEGG_pathways.csv",sep="_"))
}

```





# References: 
1. Choi, J. et al. “Haemopedia RNA-Seq: A Database of Gene Expression during Haematopoiesis in Mice and Humans.” Nucleic Acids Research 47, no. D1 (January 8, 2019): D780–85. https://doi.org/10.1093/nar/gky1020.

2. Ritchie, Matthew E. et al. “Limma Powers Differential Expression Analyses for RNA-Sequencing and Microarray Studies.” Nucleic Acids Research 43, no. 7 (April 20, 2015): e47–e47. https://doi.org/10.1093/nar/gkv007.


